src/spi_async.rs

```diff
@@ -0,0 +1,554 @@
+use crate::interrupt::{IsrCriticalSection, IsrCriticalSectionGuard};
+use crate::spi::{config, to_spi_err, Lock, SpiDriver, SpiError};
+use core::borrow::BorrowMut;
+use core::cell::UnsafeCell;
+use core::cmp::max;
+use core::cmp::Ordering;
+use core::future::Future;
+use core::iter::Zip;
+use core::marker::PhantomPinned;
+use core::ops::{Deref, DerefMut};
+use core::pin::Pin;
+use core::slice::Chunks;
+use core::slice::ChunksMut;
+use core::task::{Context, Poll, Waker};
+use embedded_hal::spi::ErrorType;
+use embedded_hal_async::spi::{SpiBus, SpiBusFlush, SpiBusRead, SpiBusWrite};
+use esp_idf_sys::*;
+
+pub struct SpiBusDriver<T> {
+    _driver: T,
+    handle: spi_device_handle_t,
+    _lock: Lock,
+    trans_len: usize,
+}
+
+// TODO: IRAM
+unsafe extern "C" fn post_transaction(transaction: *mut spi_transaction_t) {
+    let transaction = unsafe { transaction.as_mut() }.unwrap();
+
+    let waker_ptr = transaction.user as *const IsrMutex<Option<Waker>>;
+    let waker_mutex = unsafe { waker_ptr.as_ref() }.unwrap();
+
+    // Consume waker
+    let mut lock = waker_mutex.lock();
+    if let Some(waker) = lock.take() {
+        waker.wake();
+    }
+}
+
+pub struct SpiFuture<'a, T, Op: Operation> {
+    driver: &'a mut SpiBusDriver<T>, // not structurally pinned
+    transaction: spi_transaction_t,  // structurally pinned
+    state: State<Op>,                // not structurally pinned
+    _pin: PhantomPinned,
+}
+
+enum State<Op: Operation> {
+    Processing(
+        Op,                      // not structurally pinned
+        IsrMutex<Option<Waker>>, // structurally pinned
+    ),
+    Done(Result<(), SpiError>), // not structurally pinned
+}
+
+impl<'a, T, Op: Operation> Drop for SpiFuture<'a, T, Op> {
+    fn drop(&mut self) {
+        // If the future is dropped before the transaction is complete, we need to block until it
+        // is done to prevent undefined behaviour.
+        if let State::Processing(_, _) = self.state {
+            let mut ret_trans: *mut spi_transaction_t = core::ptr::null_mut();
+            esp!(unsafe {
+                spi_device_get_trans_result(self.driver.handle, &mut ret_trans as *mut _, 0)
```

Comment:
Would that even work? Especially with DMA? I mean, is there even a way to tell the C driver to immediately abort the operation so that we can reclaim our buffer? With `spi_device_get_trans_result`, really?

Reply:
`spi_device_get_trans_result` blocks until the first transaction in the queue completes, so it'll work for the current state of things.

I'm not aware of an abort API. Best we can do is use smaller transactions to block less.

Reply:
OK. I don't like this, but Rust does not have an async drop (yet).
And... in normal circumstances, user would be very unlikely to drop the SPI send/receive future on purpose. That would rather indicate a bug in his code. As in using `select!` where he should not. In fact, shall we even introduce a `warn!` log in the `drop` method if the state is `Processing`?

Reply:
Sure, I can add a `warn!` log.